### PR TITLE
lib: smf: add NULL check in smf_execute_all_exit_actions

### PR DESCRIPTION
**Pull request by ccli8**
Like in `smf_execute_all_entry_actions()`, add NULL check in `smf_execute_all_exit_actions()`.

### CODE REVIEW

**Review by keith-zephyr — DISMISSED** [review]
Looks good to me. It would be appreciated if you could extend one of the existing test cases to verify this behavior.

**Comment by keith-zephyr** [comment]
Tagging @glenn-andrews for your visibility.  If you want to nominate yourself as a collaborator on the SMF, you will automatically get added to PRs that touch the SMF.

**Comment by glenn-andrews** [comment]
What was the situation where this error occurred?

`topmost` should be a valid state, as should `ctx->current` and all parents between current and topmost.

We may have a deeper error with determining `topmost`?

**Comment by glenn-andrews** [comment]
> Tagging @glenn-andrews for your visibility. If you want to nominate yourself as a collaborator on the SMF, you will automatically get added to PRs that touch the SMF.

Done: https://github.com/zephyrproject-rtos/zephyr/pull/73729

**Comment by ccli8** [comment]
Per my check, `topmost` becomes invalid because `ctx->executing` is not updated in `smf_execute_all_entry_actions`. I add one more commit to fix it.

**Comment by ccli8** [comment]
addressed `tests/lib/smf` test failure

**Comment by glenn-andrews** [comment]
> Per my check, `topmost` becomes invalid because `ctx->executing` is not updated in `smf_execute_all_entry_actions`. I add one more commit to fix it.

Good find. With this fix, is the null check still needed?

**Comment by ccli8** [comment]
> Good find. With this fix, is the null check still needed?

It becomes unnecessary. I keep it there because it does no harm. If concerned, I can remove the null check.

File: lib/smf/smf.c [hunk 1]
```diff
@@ -98,12 +98,12 @@ static bool smf_execute_all_entry_actions(struct smf_ctx *const ctx,
 	for (const struct smf_state *to_execute = get_child_of(new_state, topmost);
 	     to_execute != NULL && to_execute != new_state;
 	     to_execute = get_child_of(new_state, to_execute)) {
+		/* Keep track of the executing entry action in case it calls
+		 * smf_set_State()
```
Comment (keith-zephyr) [inline lib/smf/smf.c:102]:
```suggestion
		 * smf_set_state()
```

**Comment by glenn-andrews** [comment]
> > Good find. With this fix, is the null check still needed?
> 
> It becomes unnecessary. I keep it there because it does no harm. If concerned, I can remove the null check.

It adds another check to an already-long if statement, and gets called for every parent. But that's still trivial compared to the CPU cycles we waste trying to find the LCA before this function gets called.

File: lib/smf/smf.c (hunk 1 above)
Comment (ccli8) [reply lib/smf/smf.c]: Fixed

**Comment by ccli8** [comment]
> Feel free to leave it in.

I keep the null check there if no other concern.

**Review by glenn-andrews — APPROVED** [review]
Looks good to me

**Review by keith-zephyr — APPROVED** [review]

**Review by sambhurst — APPROVED** [review]

**Pull request merged**